src/components/RichText.ts

```diff
@@ -0,0 +1,178 @@
+
+interface RichTextProps {
+    className?: string;
+    editorMode: EditorMode;
+    onChange?: (value: string) => void;
+    readOnly: boolean;
+    readOnlyStyle: string;
+    hasContext: boolean;
+    style?: object;
+    value: string;
+    theme: "bubble" | "snow";
+    customOptions?: Array<{ option: string }>;
+}
+
+class RichText extends Component<RichTextProps, {}> {
+    private quillNode?: HTMLElement;
+    private quill?: Quill.Quill;
+    private static quillOptions: { [key: string]: any } = {
+        align: { align: [] },
+        blockQuote: "blockquote",
+        bold: "bold",
+        bulletList: { list: "bullet" },
+        codeBlock: "code-block",
+        direction: { direction: "rtl" },
+        fillColor: { background: [] },
+        headers: { header: [ 1, 2, 3, 4, 5, 6, false ] },
+        indent: { indent: "-1" },
+        italic: "italic",
+        orderedList: { list: "ordered" },
+        outdent: { indent: "+1" },
+        strike: "strike",
+        subScript: { script: "sub" },
+        superScript: { script: "super" },
+        textColor: { color: [] },
+        underline: "underline"
+    };
+
+    constructor(props: RichTextProps) {
+        super(props);
+
+        this.state = { text: props.value };
+        this.handleSelectionChange = this.handleSelectionChange.bind(this);
+        this.setQuillNode = this.setQuillNode.bind(this);
+    }
+
+    render() {
+        return DOM.div({
+                className: classNames("widget-rich-text", this.props.className, {
+                    "no-context": !this.props.hasContext,
+                    "read-only": this.props.readOnly && this.props.readOnlyStyle === "text"
+                }),
+                style: this.props.style
+            },
+            DOM.div({ className: "widget-rich-text-quill", ref: this.setQuillNode })
+        );
+    }
+
+    componentDidMount() {
+        this.renderEditor(this.props);
+    }
+
+    componentWillReceiveProps(nextProps: RichTextProps) {
+        this.updateEditor(nextProps);
+    }
+
+    componentWillUnmount() {
+        this.handleSelectionChange();
+    }
+
+    private setQuillNode(node: HTMLElement) {
+        this.quillNode = node;
+    }
+
+    private renderEditor(props: RichTextProps) {
+        if (this.quillNode && !this.quill) {
+            this.quill = new Quill(this.quillNode, {
+                modules: this.getEditorModules(),
+                theme: props.theme
+            });
+
+            this.quill.on("selection-change", this.handleSelectionChange);
+
+        }
+    }
+
+    private handleSelectionChange() {
+        if (this.quill && !this.quill.hasFocus() && this.props.onChange) {
+            this.props.onChange((this.quill as any).container.firstChild.innerHTML);
+        }
+    }
+
+    private updateEditor(props: RichTextProps) {
+        if (this.quill) {
+            this.quill.enable(!props.readOnly);
+            this.quill.clipboard.dangerouslyPasteHTML(props.value);
```

Comment:
Applied but still needs some testing... selection seems to always be null locally.